internal/controller/purge_controller.go

```diff
@@ -61,60 +61,61 @@ func (r *PurgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	kyma := &v1beta2.Kyma{}
 	if err := r.Get(ctx, req.NamespacedName, kyma); err != nil {
 		if util.IsNotFound(err) {
-			logger.V(log.DebugLevel).Info(fmt.Sprintf("Kyma %s not found, probably already deleted",
-				req.NamespacedName))
-			return ctrl.Result{Requeue: false}, nil
+			logger.V(log.DebugLevel).Info(fmt.Sprintf("Kyma %s not found, probably already deleted", req.NamespacedName))
+			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, fmt.Errorf("purgeController: %w", err)
+		return ctrl.Result{}, fmt.Errorf("failed getting Kyma %s: %w", req.NamespacedName, err)
 	}
 
 	if kyma.DeletionTimestamp.IsZero() {
-		err := r.ensurePurgeFinalizer(ctx, kyma)
-		if err != nil {
+		if err := r.ensurePurgeFinalizer(ctx, kyma); err != nil {
+			logger.V(log.DebugLevel).Info(fmt.Sprintf("Failed setting purge finalizer for Kyma %s: %s", req.NamespacedName, err))
+			r.raiseSettingPurgeFinalizerFailedEvent(kyma, err)
 			return ctrl.Result{}, err
 		}
-		return ctrl.Result{Requeue: false}, nil
+		return ctrl.Result{}, nil
 	}
 
-	requeueAfter := r.calculateRequeueAfterTime(ctx, kyma)
-	if requeueAfter != 0 {
+	if requeueAfter := r.calculateRequeueAfterTime(kyma); requeueAfter != 0 {
+		logger.Info(fmt.Sprintf("Purge reconciliation for Kyma %s will be requeued after %s", req.NamespacedName, requeueAfter))
 		return ctrl.Result{RequeueAfter: requeueAfter}, nil
 	}
 
 	start := time.Now()
 	remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKeyFromObject(kyma))
-	if util.IsNotFound(err) {
-		if err := r.dropPurgeFinalizer(ctx, kyma); err != nil {
-			r.Metrics.UpdatePurgeError(ctx, kyma, metrics.ErrPurgeFinalizerRemoval)
-			return ctrl.Result{}, err
-		}
-		return ctrl.Result{Requeue: true}, nil
-	}
 	if err != nil {
-		return ctrl.Result{}, err
+		if util.IsNotFound(err) {
+			if err := r.dropPurgeFinalizer(ctx, kyma); err != nil {
+				r.raiseRemovingPurgeFinalizerFailedError(ctx, kyma, err)
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed getting remote client for Kyma %s: %w", req.NamespacedName, err)
 	}
 
 	r.Metrics.UpdatePurgeCount()
 	if err := r.performCleanup(ctx, remoteClient); err != nil {
-		logger.Error(err, "Purge Cleanup failed")
+		logger.Error(err, fmt.Sprintf("failed purge cleanup for Kyma %s", req.NamespacedName))
 		r.Metrics.UpdatePurgeError(ctx, kyma, metrics.ErrCleanup)
 		return ctrl.Result{}, err
 	}
 
 	if err := r.dropPurgeFinalizer(ctx, kyma); err != nil {
-		r.Metrics.UpdatePurgeError(ctx, kyma, metrics.ErrPurgeFinalizerRemoval)
+		r.raiseRemovingPurgeFinalizerFailedError(ctx, kyma, err)
 		return ctrl.Result{}, err
 	}
-	duration := time.Since(start)
-	r.Metrics.UpdatePurgeTime(duration)
+
+	r.Metrics.UpdatePurgeTime(time.Since(start))
```

Comment:
This metrics actually only record happy path, but for those kyma end up into error cleanup, we don't have any metrics reflect it, if we can improve this metrics to update purge time also include error case, it also help us identify those kyma not perform cleanup within expected time.

Reply:
Apologies, not sure if I can fully follow you. For the unhappy path, we do have:

- log + metric + event, if dropping the purge finalizer from Kyma fails:
  - [here](https://github.com/kyma-project/lifecycle-manager/blob/371c5e32cd551a507d41da74d9262a3b850081a1/internal/controller/purge_controller.go#L89), and
  - [here](https://github.com/kyma-project/lifecycle-manager/blob/371c5e32cd551a507d41da74d9262a3b850081a1/internal/controller/purge_controller.go#L106)
- metric, if the cleanup failed
  - [here](https://github.com/kyma-project/lifecycle-manager/blob/371c5e32cd551a507d41da74d9262a3b850081a1/internal/controller/purge_controller.go#L101)

The metric is Kyma specific, so we can derive the time of trying the purge from the time the counter was increased, e.g., here:
![image](https://github.com/kyma-project/lifecycle-manager/assets/22994291/957b5a93-52ee-47d0-9c49-8f353d2838c1)

Reply:
Added the reset of the value here: https://github.com/kyma-project/lifecycle-manager/pull/1392/commits/d668ab42f9cd87d79808e9af1ea649532acf99c8